src/silo/query_engine/actions/action.cpp

```diff
@@ -140,6 +150,45 @@ void from_json(const nlohmann::json& json, OrderByField& field) {
    field = {field_name, order_string == "ascending"};
 }
 
+std::optional<uint32_t> parseLimit(const nlohmann::json& json) {
+   CHECK_SILO_QUERY(
+      !json.contains("limit") || json["limit"].is_number_unsigned(),
+      "If the action contains a limit, it must be a non-negative number"
+   )
+   return json.contains("limit") ? std::optional<uint32_t>(json["limit"].get<uint32_t>())
+                                 : std::nullopt;
+}
+
+std::optional<uint32_t> parseOffset(const nlohmann::json& json) {
+   CHECK_SILO_QUERY(
+      !json.contains("offset") || json["offset"].is_number_unsigned(),
+      "If the action contains an offset, it must be a non-negative number"
+   )
+   return json.contains("offset") ? std::optional<uint32_t>(json["offset"].get<uint32_t>())
+                                  : std::nullopt;
+}
+
+std::optional<uint32_t> parseRandomizeSeed(const nlohmann::json& json) {
+   if (json.contains("randomize")) {
```

Comment:
This is a nice case where early returns improve readability.

```suggestion
   if (!json.contains("randomize")) {
      return std::nullopt;
   }
```